standalone-metastore/metastore-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalogAdapter.java

```diff
@@ -0,0 +1,622 @@
+
+      this.requiredLength = parts.size();
+      this.requirements = requirementsBuilder.build();
+      this.variables = variablesBuilder.build();
+    }
+
+    private boolean matches(HTTPMethod requestMethod, List<String> requestPath) {
+      return method == requestMethod &&
+          requiredLength == requestPath.size() &&
+          requirements.entrySet().stream()
+              .allMatch(
+                  requirement ->
+                      requirement
+                          .getValue()
+                          .equalsIgnoreCase(requestPath.get(requirement.getKey())));
+    }
+
+    private Map<String, String> variables(List<String> requestPath) {
+      ImmutableMap.Builder<String, String> vars = ImmutableMap.builder();
+      variables.forEach((key, value) -> vars.put(value, requestPath.get(key)));
+      return vars.build();
+    }
+
+    public static Pair<Route, Map<String, String>> from(HTTPMethod method, String path) {
+      List<String> parts = SLASH.splitToList(path);
+      for (Route candidate : Route.values()) {
+        if (candidate.matches(method, parts)) {
+          return Pair.of(candidate, candidate.variables(parts));
+        }
+      }
+
+      return null;
+    }
+
+    public Class<? extends RESTRequest> requestClass() {
+      return requestClass;
+    }
+
+    public Class<? extends RESTResponse> responseClass() {
+      return responseClass;
+    }
+  }
+
+  /**
+   * @param route a route/api-call name
+   * @return the metric counter name for the api-call
+   */
+  static String hmsCatalogMetricCount(String route) {
+    return HMS_METRIC_PREFIX + route.toLowerCase() + ".count";
+  }
+
+  /**
+   * @param apis an optional list of known api call names
+   * @return the list of metric names for the HMSCatalog class
+   */
+  public static List<String> getMetricNames(String... apis) {
+    final List<Route> routes;
+    if (apis != null && apis.length > 0) {
+      routes = Arrays.stream(apis)
+          .map(HMSCatalogAdapter.Route::byName)
+          .filter(Objects::nonNull)
+          .collect(Collectors.toList());
+    } else {
+      routes = Arrays.asList(HMSCatalogAdapter.Route.values());
+    }
+    final List<String> metricNames = new ArrayList<>(routes.size());
+    for (HMSCatalogAdapter.Route route : routes) {
+      metricNames.add(hmsCatalogMetricCount(route.name()));
+    }
+    return metricNames;
+  }
+
+  @SuppressWarnings("MethodLength")
+   <T extends RESTResponse> T handleRequest(
+      Route route, Map<String, String> vars, Object body, Class<T> responseType) {
+    // update HMS catalog route counter metric
+    final String metricName = hmsCatalogMetricCount(route.name());
+    Counter counter = Metrics.getOrCreateCounter(metricName);
+    if (counter != null) {
+      counter.inc();
+    }
+    switch (route) {
+      case TOKENS: {
+        @SuppressWarnings("unchecked")
+        Map<String, String> request = (Map<String, String>) castRequest(Map.class, body);
```

Comment:
This switch contains 17 cases and even switch inside the case, which is really hard to follow what is happening here. Instead of adding the implementation inside the cases please transform it like:

switch (route) {
 case TOKENS:
  handleTokensRequest(...);
  case CONFIG:
   handleConfigRequest(...); 
  case LIST_NAMESPACES
   handleListNamespacesRequest(...);
  ...
}

This way it is much easier to understand, modify, maintain.

Reply:
Will do.